fdbserver/TLogServer.actor.cpp

```diff
@@ -1805,73 +1805,76 @@ Future<Void> tLogPeekMessages(PromiseType replyPromise,
+			wait(waitForMessagesForTag(logData, reqTag, reqBegin, SERVER_KNOBS->BLOCKING_PEEK_TIMEOUT));
+			double latency = now() - startTime;
+			if (logData->blockingPeekLatencies.find(reqTag) == logData->blockingPeekLatencies.end()) {
+				UID ssID = nondeterministicRandom()->randomUniqueID();
+				std::string s = "BlockingPeekLatencies-" + reqTag.toString();
+				logData->blockingPeekLatencies.try_emplace(
+				    reqTag, s, ssID, SERVER_KNOBS->LATENCY_METRICS_LOGGING_INTERVAL, SERVER_KNOBS->LATENCY_SAMPLE_SIZE);
+			}
+			LatencySample& sample = logData->blockingPeekLatencies.at(reqTag);
+			sample.addMeasurement(latency);
+			poppedVer = poppedVersion(logData, reqTag);
+		}
 
-		if (reqSequence.present()) {
-			auto& trackerData = logData->peekTracker[peekId];
-			auto& sequenceData = trackerData.sequence_version[sequence + 1];
-			trackerData.lastUpdate = now();
-			if (trackerData.sequence_version.size() && sequence + 1 < trackerData.sequence_version.begin()->first) {
-				replyPromise.sendError(operation_obsolete());
-				if (!sequenceData.isSet())
-					sequenceData.sendError(operation_obsolete());
-				return Void();
-			}
-			if (sequenceData.isSet()) {
-				if (sequenceData.getFuture().get().first != rep.end) {
-					TEST(true); // tlog peek second attempt ended at a different version
+		DisabledTraceEvent("TLogPeekMessages1", self->dbgid)
+		    .detail("LogId", logData->logId)
+		    .detail("Tag", reqTag.toString())
+		    .detail("ReqBegin", reqBegin)
+		    .detail("PoppedVer", poppedVer);
+
+		if (poppedVer > reqBegin) {
+			TLogPeekReply rep;
+			rep.maxKnownVersion = logData->version.get();
+			rep.minKnownCommittedVersion = logData->minKnownCommittedVersion;
+			rep.popped = poppedVer;
+			rep.end = poppedVer;
+			rep.onlySpilled = false;
+
+			if (reqSequence.present()) {
+				auto& trackerData = logData->peekTracker[peekId];
+				auto& sequenceData = trackerData.sequence_version[sequence + 1];
+				trackerData.lastUpdate = now();
+				if (trackerData.sequence_version.size() && sequence + 1 < trackerData.sequence_version.begin()->first) {
 					replyPromise.sendError(operation_obsolete());
+					if (!sequenceData.isSet())
+						sequenceData.sendError(operation_obsolete());
 					return Void();
 				}
-			} else {
-				sequenceData.send(std::make_pair(rep.end, rep.onlySpilled));
+				if (sequenceData.isSet()) {
+					if (sequenceData.getFuture().get().first != rep.end) {
+						TEST(true); // tlog peek second attempt ended at a different version
+						replyPromise.sendError(operation_obsolete());
+						return Void();
+					}
+				} else {
+					sequenceData.send(std::make_pair(rep.end, rep.onlySpilled));
+				}
+				rep.begin = reqBegin;
 			}
-			rep.begin = reqBegin;
-		}
 
-		replyPromise.send(rep);
-		return Void();
-	}
+			replyPromise.send(rep);
+			return Void();
+		}
 
-	state Version endVersion;
-	state bool onlySpilled;
+		ASSERT_WE_THINK(reqBegin >= poppedVersion(logData, reqTag));
```

Comment:
Do we want to use `ASSERT` instead? `ASSERT_WE_THINK` is noop in release build. Or, if this conditions triggers, return an error instead?

Reply:
The idea here is that we compute latest popped version using `poppedVersion()` instead of reading from (cached) poppedVer. So I want to avoid this computation in prod.

Also, I feel this should be an invarant rather than an error, so it's better to be caught in simulation.

Reply:
`poppedVersion()` only does a map lookup, so it should be cheap. I am fine to keep as is.

Reply:
I think it's also true when `ENABLE_VERSION_VECTOR` is enabled since we lookup and set `poppedVer` at the end of that block. I'm in favor of just changing this to the `ASSERT` then if this is inexpensive, otherwise you can change it to the formulation I described above.

Reply:
All, good suggestion. I change it to `ASSERT`.